clients/drcachesim/tests/cache_replacement_policy_unit_test.cpp

```diff
@@ -105,9 +140,91 @@ unit_test_cache_lru_four_way()
     cache_lru_test.access_and_check_lru(ADDRESS_E, 2); // A E c D
 }
 
+void
+unit_test_cache_fifo_four_way()
+{
+    cache_fifo_test_t cache_fifo_test;
+    cache_fifo_test.initialize_cache(/*associativity=*/4, /*line_size=*/32,
+                                     /*total_size=*/256);
+    const addr_t ADDRESS_A = 0;
+    const addr_t ADDRESS_B = 64;
+    const addr_t ADDRESS_C = 128;
+    const addr_t ADDRESS_D = 192;
+    const addr_t ADDRESS_E = 72;
+
+    assert(cache_fifo_test.get_block_index(ADDRESS_A) ==
+           cache_fifo_test.get_block_index(ADDRESS_B));
+    assert(cache_fifo_test.get_block_index(ADDRESS_B) ==
+           cache_fifo_test.get_block_index(ADDRESS_C));
+    assert(cache_fifo_test.get_block_index(ADDRESS_C) ==
+           cache_fifo_test.get_block_index(ADDRESS_D));
+    assert(cache_fifo_test.get_block_index(ADDRESS_D) ==
+           cache_fifo_test.get_block_index(ADDRESS_E));
+
+    // Lower-case letter shows the way that is to be replaced after the access.
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); // A x X X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_B, 2); // A B x X
+    cache_fifo_test.access_and_check_fifo(ADDRESS_C, 3); // A B C x
+    cache_fifo_test.access_and_check_fifo(ADDRESS_D, 0); // a B C D
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); // A b C D
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); // A b C D
+    cache_fifo_test.access_and_check_fifo(ADDRESS_A, 1); // A b C D
+    cache_fifo_test.access_and_check_fifo(ADDRESS_E, 2); // A B c D
```

Comment:
This is not correct: A came into the cache first and so should be the first one evicted.

Reply:
Done.